Updates to scheduling objects to support dynamic custom resources

romilbhardwaj opened the pull request:
## What do these changes do?

These changes add support to scheduling objects (ResourceSet, ResourceIdSet) to support adding, updating and removing resources at runtime. This is a split of the larger PR #3742.

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -379,6 +452,18 @@ class SchedulingResources {
+
+  /// \brief Delete resource from total, available and load resources.
+  ///
+  /// \param
```
fill out the documentation

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -72,6 +73,11 @@ bool ResourceSet::AddResource(const std::string &resource_name, double capacity)
 
+bool ResourceSet::DeleteResource(const std::string &resource_name) {
+  resource_capacity_.erase(resource_name);
+  return true;
```
no point in returning this

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -92,6 +98,29 @@ bool ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
+  for (const auto &resource_pair : other.GetResourceMap()) {
+    const std::string &resource_label = resource_pair.first;
+    const double &resource_capacity = resource_pair.second;
+    RAY_CHECK(resource_capacity_.count(resource_label) == 1)
```
Is `resource_capacity` guaranteed to be positive? If so, include a check. If not, then it's ok for this check to fail.

romilbhardwaj replied:
This function was an overload of `SubtractResourcesStrict(const ResourceSet &other)`. I have merged the two functions into one `SubtractResourcesStrict(const ResourceSet &other, bool delete_zero_capacity)` and replaced the old calls to `SubtractResourcesStrict(res_set)` with   `SubtractResourcesStrict(res_set, false)` in a new commit.

This check existed in the other function too so I didn't touch it. Should we remove it? I think it's a good sanity check to have to before subtracting resources.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -170,16 +261,34 @@ ResourceIds::ResourceIds(double resource_quantity) {
+ResourceIds::ResourceIds(const std::vector<std::pair<int64_t, double>> &fractional_ids){
+  fractional_ids_ = fractional_ids;
+  // Find the max id and set it to greatest_id_;
+  const auto p = std::minmax_element(fractional_ids_.begin(), fractional_ids_.end());
```
How is `max` defined for pairs?

romilbhardwaj replied:
We can use `std::max_element` but will have to define a custom comparator to compare pairs. `std::minmax_element` has a default comparator for pairs which evaluates the second value of the pair to find the max and min.

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -227,6 +279,14 @@ class ResourceIds {
+  /// A double to track the greatest id of the resource - this needs to be maintained to
+  /// ensure unique ids are used when creating new resources. This keeps incrementing,
+  /// and resources may have gaps in between. TODO(romilb): Is this a safe assumption?
+  double greatest_id_;
```
should be `int64_t`, right?

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -227,6 +279,14 @@ class ResourceIds {
   std::vector<std::pair<int64_t, double>> fractional_ids_;
+  /// A double to track the greatest id of the resource - this needs to be maintained to
+  /// ensure unique ids are used when creating new resources. This keeps incrementing,
+  /// and resources may have gaps in between. TODO(romilb): Is this a safe assumption?
```
It'd be preferable not to have gaps. How can gaps arise?

romilbhardwaj replied:
Gaps can arise when resources are deleted (through a call to `ResourceIds::DecreaseCapacity`). On deletion, we cannot simply decrement `greatest_id_` because the first available resource in the `whole_ids_` vector gets removed (which may not necessarily be the resource with the largest id).

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -170,16 +261,34 @@ ResourceIds::ResourceIds(double resource_quantity) {
     whole_ids_.push_back(i);
   }
+  greatest_id_ = resource_quantity;
+  total_capacity_ = TotalQuantity();
```
If we're adding an extra `total_capacity_` field, then do we need the `TotalQuantity` method? Can't we just have `TotalQuantity` return `total_capacity_`?

romilbhardwaj replied:
I think the name `TotalQuantity()` is a bit misleading, since it returns the number of resources currently in the `whole_ids_` vector. The `whole_ids_` vector size keeps changing when resources are acquired or released, thus `TotalQuantity()` reflects the available resources. Thus I added `total_capacity_` to keep a track of the total count of resources in `ResourceIds`.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -505,14 +686,44 @@ const ResourceSet &SchedulingResources::GetLoadResources() const {
 // Return specified resources back to SchedulingResources.
 bool SchedulingResources::Release(const ResourceSet &resources) {
-  return resources_available_.AddResourcesStrict(resources);
+  bool avail_status = resources_available_.AddResourcesCapacityConstrained(resources, resources_total_);
```
avoid using abbreviations everywhere, so `avail` should be `available`.

Also in this case, it's not clear what `avail_status` means. Maybe choose a better name?

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -102,6 +115,27 @@ class ResourceSet {
+  ///
+  /// \param new_res_set: The new resource set to compare with.
+  /// \return The ResourceSet of updated values
+  ResourceSet FindUpdatedResources(const ResourceSet &new_res_set) const;
```
`res` -> `resource` everywhere

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -372,9 +530,15 @@ void ResourceIdSet::Release(const ResourceIdSet &resource_id_set) {
-      // This should not happen when Release is called on resources that were obtained
-      // through a corresponding call to Acquire.
-      available_resources_[resource_name] = resource_ids;
+      if (strict==false){
```
`if (!strict) {`

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -264,8 +324,9 @@ class ResourceIdSet {
   /// \brief Return a set of resource IDs.
   ///
   /// \param resource_id_set The resource IDs to return.
+  /// \param strict If set to to true, creates any resources that do not already exist in the ResourceIdSet. Else ignores any new resources.
```
Why would we ever want to ignore new resources?

romilbhardwaj replied:
The resource might have been deleted while it was acquired by some task, and thus a release should not add them back to `available_resources_`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13212/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13215/
Test FAILed.

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -379,6 +444,19 @@ class SchedulingResources {
+  ///
+  /// \param resource_name: Name of the resource to be deleted.
+  /// \return Void.
+  void DeleteResource(std::string resource_name);
```
`const std::string &resource_name`

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -379,6 +444,19 @@ class SchedulingResources {
+  /// \param resource_name: Name of the resource to be modified
+  /// \param capacity: New capacity of the resource.
+  /// \return Void.
+  void UpdateResource(std::string resource_name, double capacity);
```
`const std::string &resource_name`

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -278,6 +331,18 @@ class ResourceIdSet {
+  /// \brief Deletes a resource in the ResourceIdSet. This does not raise an exception, just deletes the resource. Tasks with acquired resources keep running.
+  ///
+  /// \param resource_name the name of the resource to delete
+  void DeleteResource(const std::string resource_name);
```
`const std::string &resource_name`

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -278,6 +331,18 @@ class ResourceIdSet {
+  ///
+  /// \param resource_name the name of the resource to create/update
+  /// \param capacity capacity of the resource being added
+  void CreateResource(const std::string resource_name, const double capacity);
```
`const std::string &resource_name`

robertnishihara commented:
@romilbhardwaj the tests seem to be failing. I can reproduce this locally by running e.g.,

```
python -m pytest -v --durations=10 python/ray/tune/tests/test_commands.py::test_ls
```

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -227,6 +271,14 @@ class ResourceIds {
   std::vector<std::pair<int64_t, double>> fractional_ids_;
+  /// int to track the greatest id of the resource - this needs to be maintained to
+  /// ensure unique ids are used when creating new resources. This keeps incrementing,
+  /// and resources may have gaps in between.
```
I don't think it's acceptable to have gaps in the IDs. That will give the user an inconsistent experience (e.g., what IDs they get will depend on when things get scheduled). Instead, how about we simply don't provide IDs for dynamic resources?

raulchen replied:
I prefer not providing IDs as well.

romilbhardwaj commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -278,6 +331,18 @@ class ResourceIdSet {
+  ///
+  /// \param resource_name the name of the resource to create/update
+  /// \param capacity capacity of the resource being added
+  void CreateResource(const std::string resource_name, const double capacity);
```
I agree, renaming all Create/Add methods to `AddOrUpdateResource`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13237/
Test FAILed.

romilbhardwaj commented:
Thanks for the comments, @robertnishihara and @raulchen. I had a closer look at `SubtractResourcesStrict` and realized the `delete_zero_capacity` flag wasn't needed for dynamic resources anymore and infact was related to the bug which was causing the tune tests to fail . I've removed it and reverted `SubtractResourcesStrict` to it's original implementation.

raulchen commented:
Thanks, I'll take another look tomorrow. Note, the linter is failing. https://travis-ci.com/ray-project/ray/jobs/187695736

romilbhardwaj commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -505,12 +669,42 @@ const ResourceSet &SchedulingResources::GetLoadResources() const {
 // Return specified resources back to SchedulingResources.
 bool SchedulingResources::Release(const ResourceSet &resources) {
-  return resources_available_.AddResourcesStrict(resources);
+  bool release_status = resources_available_.AddResourcesCapacityConstrained(resources, resources_total_);
```
We need to constrain `Release` now because resources may get removed while they are acquired by a task, and a subsequent release would incorrectly re-add those resources to `resources_available_`. 

I have removed `AddResourcesStrict`.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13271/
Test PASSed.

robertnishihara commented:
`greatest_id_` should no longer be necessary. Let's remove that and get rid of the concept of IDs for dynamic resources.

romilbhardwaj commented:
> `greatest_id_` should no longer be necessary. Let's remove that and get rid of the concept of IDs for dynamic resources.

Okay, but we would still need to allocate the newly created resources some id which can be pushed in to `ResourceIds::whole_ids_`. In the latest commit, I assign `-1` as the id to all dynamically updated resources - would that be okay?

This would be the new behavior when the user calls `ray.get_resource_ids`:
```
>>> ray.experimental.create_resource("a",1)	# This creates 1 resource with id 0 
>>> ray.experimental.create_resource("a",5)	# This creates 4 more resources with ids -1.
>>> @ray.remote(resources={"a":5})
... def f():  
...  return ray.get_resource_ids()
... 
>>> print(ray.get(f.remote()))
{'a': [(-1, 1.0), (-1, 1.0), (-1, 1.0), (-1, 1.0), (0, 1.0)], 'CPU': [(7, 1.0)]}
```

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13305/
Test PASSed.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -170,16 +225,27 @@ ResourceIds::ResourceIds(double resource_quantity) {
+ResourceIds::ResourceIds(const std::vector<std::pair<int64_t, double>> &fractional_ids) {
+  fractional_ids_ = fractional_ids;
+  total_capacity_ = TotalQuantity();
+  decrement_backlog_ = 0;
```
Should these be initialized like

```cpp
fractional_ids_(fractional_ids),
```

etc?

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -302,6 +377,47 @@ std::string ResourceIds::ToString() const {
+void ResourceIds::IncreaseCapacity(double increment_quantity) {
+  for (int i = 0; i < increment_quantity; i++) {
+    whole_ids_.push_back(-1);     // Dynamic resources are assigned resource id -1.
+  }
```
Do we want to take stuff off the decrement backlog in this case?

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -302,6 +377,47 @@ std::string ResourceIds::ToString() const {
   return return_string;
 }
 
+void ResourceIds::UpdateCapacity(double new_capacity) {
```
If this is called, consider checking that this is a dynamic resource.

robertnishihara replied:
Also, `new_capacity` will must be an integer, right?

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -215,6 +236,26 @@ class ResourceIds {
+  ///
+  /// \param increment_quantity A double of how many unit resources to add.
+  /// \return Void.
+  void IncreaseCapacity(double increment_quantity);
```
`IncreaseCapacity` and `DecreaseCapacity` should be private, and `UpdateCapacity` is the only public one.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -361,7 +477,7 @@ ResourceIdSet ResourceIdSet::Acquire(const ResourceSet &resource_set) {
 }
 
-void ResourceIdSet::Release(const ResourceIdSet &resource_id_set) {
+void ResourceIdSet::Release(const ResourceIdSet &resource_id_set, bool strict) {
```
`strict` -> `allow_new_resources` (which will require flipping the if statement)

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -505,14 +645,44 @@ const ResourceSet &SchedulingResources::GetLoadResources() const {
   return resources_available_.SubtractResourcesStrict(resources);
 }
 
+void SchedulingResources::UpdateResource(std::string &resource_name, double capacity) {
```
`const std::string &resource_name`

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -505,14 +645,44 @@ const ResourceSet &SchedulingResources::GetLoadResources() const {
+  }
+}
+
+void SchedulingResources::DeleteResource(std::string &resource_name) {
```
const

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -92,14 +97,27 @@ bool ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
-    RAY_CHECK(resource_capacity_.count(resource_label) != 0);
-    resource_capacity_[resource_label] += resource_capacity;
+    const std::string &to_add_resource_label = resource_pair.first;
+    const double &to_add_resource_capacity = resource_pair.second;
```
```suggestion
    double to_add_resource_capacity = resource_pair.second;
```

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -92,14 +97,27 @@ bool ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
+// capacity from the total_resource set
+bool ResourceSet::AddResourcesCapacityConstrained(const ResourceSet &other,
+                                                  const ResourceSet &total_resources) {
+  const std::unordered_map<std::string, double> total_resource_map =
```
```suggestion
  const std::unordered_map<std::string, double> &total_resource_map =
```

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -92,14 +97,27 @@ bool ResourceSet::SubtractResourcesStrict(const ResourceSet &other) {
+                   total_capacity);
+    } else {
+      // Resource does not exist in the total map, it probably got deleted from the total.
+      // Don't panic, do nothing and simply continue.
```
Is it worth logging a warning here?

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -160,6 +178,43 @@ const std::unordered_map<std::string, double> &ResourceSet::GetResourceMap() con
+  ResourceSet updated_resource_set;
+  for (const auto &resource_pair : new_resource_set.GetResourceMap()) {
+    const std::string &resource_label = resource_pair.first;
+    const double &new_resource_capacity = resource_pair.second;
```
```suggestion
    double new_resource_capacity = resource_pair.second;
```

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -160,6 +178,43 @@ const std::unordered_map<std::string, double> &ResourceSet::GetResourceMap() con
+  auto &new_resource_map = new_resource_set.GetResourceMap();
+  for (const auto &resource_pair : resource_capacity_) {
+    const std::string &resource_label = resource_pair.first;
+    const double &old_resource_capacity = resource_pair.second;
```
```suggestion
    double old_resource_capacity = resource_pair.second;
```

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -278,6 +327,20 @@ class ResourceIdSet {
+  ///
+  /// \param resource_name the name of the resource to create/update
+  /// \param capacity capacity of the resource being added
+  void AddOrUpdateResource(const std::string &resource_name, const double capacity);
```
```suggestion
  void AddOrUpdateResource(const std::string &resource_name, double capacity);
```

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -385,10 +508,27 @@ void ResourceIdSet::Clear() { available_resources_.clear(); }
 }
 
+void ResourceIdSet::AddOrUpdateResource(const std::string &resource_name,
+                                        const double capacity) {
```
```suggestion
                                        double capacity) {
```

robertnishihara commented:
When running `python -m pytest -s python/ray/tests/test_actor.py::test_resource_assignment`, I see the test hang sometimes (sometimes it passes). When it hangs, I see this in the raylet logs.

```
F0327 17:10:36.986322 2466321280 scheduling_resources.cc:336]  Check failed: fractional_pair_it->second <= 1 
*** Check failure stack trace: ***
*** Aborted at 1553731836 (unix time) try "date -d @1553731836" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x7fff5a64ab66) received by PID 18892 (TID 0x7fff93011380) stack trace: ***
    @     0x7fff5a808f5a _sigtramp
    @     0x7ffee68dc268 (unknown)
    @     0x7fff5a5a61ae abort
    @        0x109440569 google::logging_fail()
    @        0x10943f242 google::LogMessage::SendToLog()
    @        0x10943f945 google::LogMessage::Flush()
    @        0x10943fa22 google::LogMessage::~LogMessage()
    @        0x109439205 ray::RayLog::~RayLog()
    @        0x10937868d ray::raylet::ResourceIds::Release()
    @        0x10937a804 ray::raylet::ResourceIdSet::Release()
    @        0x10934e255 ray::raylet::NodeManager::ProcessDisconnectClientMessage()
    @        0x10934d003 ray::raylet::NodeManager::ProcessClientMessage()
    @        0x10936c85c std::__1::__function::__func<>::operator()()
    @        0x10942d85c ray::ClientConnection<>::ProcessMessage()
    @        0x109435698 boost::asio::detail::reactive_socket_recv_op<>::do_complete()
    @        0x1093306df boost::asio::detail::scheduler::do_run_one()
    @        0x109330102 boost::asio::detail::scheduler::run()
    @        0x10932537e main
    @     0x7fff5a4fa015 start
    @               0x10 (unknown)
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13332/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13334/
Test FAILed.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -302,6 +375,53 @@ std::string ResourceIds::ToString() const {
+
+void ResourceIds::IncreaseCapacity(int64_t increment_quantity) {
+  // Adjust with decrement_backlog_
+  double actual_increment_quantity = std::max(0.0, increment_quantity-decrement_backlog_);
```
Why are we turning an `int` into a `double`?

romilbhardwaj replied:
oh sorry I missed this while refactoring - fixed now

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13338/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13344/
Test PASSed.

jovany-wang commented:
@robertnishihara I rebase current master again. Could you please try to compile this?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13345/
Test FAILed.

robertnishihara commented:
Jenkins, retest this please.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-Perf-Integration-PRB/312/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13350/
Test FAILed.

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -302,6 +377,57 @@ std::string ResourceIds::ToString() const {
+void ResourceIds::UpdateCapacity(int64_t new_capacity) {
+  // Assert the new capacity is positive for sanity
+  RAY_CHECK(new_capacity >= 0);
+  int64_t capacity_delta = new_capacity - total_capacity_;
```
I'm confused about this. Is `total_capacity_` always an int? It's currently declared as a double.

romilbhardwaj replied:
`total_capacity_` is double because by definition, it should include fractional resources. I could change it to int to reflect it's current usage, but I think it's better to leave it as a double so it stays true to it's definition. Thanks for pointing this out though, I've updated this line to `int64_t capacity_delta = new_capacity - (int64_t)total_capacity_;` to truncate any fractional resources from this calculation.

robertnishihara replied:
`total_capacity_` is initialized to an integer, right? And it's only updated in `UpdateCapacity`, which requires an integer value, right? So when could it be not an integer?

Also, the line `int64_t capacity_delta = new_capacity - (int64_t)total_capacity_;` really concerns me because if `total_capacity_` has some fractional value, then we are truncating it and losing that value. Wouldn't that lead to incorrect behavior?

Also, if we do need to cast it, please right it as `int64_t capacity_delta = new_capacity - static_cast<int64_t>(total_capacity_);` to avoid C-style casts.

romilbhardwaj replied:
`total_capacity_` is not necessarily initialized as an integer - it may be initialized as a double when a `ResourceIds` object is [instantiated with fractional resources](https://github.com/romilbhardwaj/ray/blob/ce30cdbc1238d21787a0dba17504d09c93ebdc7a/src/ray/raylet/scheduling_resources.cc#L238-L248). 

I think the cast to int is required in `int64_t capacity_delta = new_capacity - (int64_t)total_capacity_;` for correct behavior. Consider the two possible cases:
1) `total_capacity_` is a whole number -> there are no fractional resources. The cast has no effect.
2) `total_capacity_` is not a whole number -> there exist fractional resources. If we do not do the explicit `int64_t` cast, the subtraction `new_capacity - total_capacity_` followed by a `int64_t` cast for `capacity_delta` would result in an incorrect result. For instance, consider new_capacity = 8 and total_capacity_ = 7.2. The value of `capacity_delta` after evaluating `int64_t capacity_delta = new_capacity - (int64_t)total_capacity_;` would be 0, whereas we would want it to be 1 (and thus the resultant capacity would be 8.2).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13352/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13351/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13359/
Test FAILed.

raulchen commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -67,25 +67,34 @@ class ResourceSet {
-  /// \return True if the resource set was added successfully. False otherwise.
-  bool AddResourcesStrict(const ResourceSet &other);
+  /// \param total_resources: Total resource set which sets upper limits on capacity for
+  /// each label. \return True if the resource set was added successfully. False
```
```suggestion
  /// each label. 
  /// \return True if the resource set was added successfully. False
```

raulchen commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -264,8 +311,10 @@ class ResourceIdSet {
-  void Release(const ResourceIdSet &resource_id_set);
+  /// \param add_new_resources If set to to true, creates any resources that do not
+  /// already exist in the ResourceIdSet. Else ignores any new resources and does not add
+  /// them back to available_resources_. \return Void.
```
```suggestion
  /// them back to available_resources_. 
  /// \return Void.
```

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13374/
Test FAILed.

robertnishihara commented on src/ray/raylet/scheduling_resources.h:
```diff
@@ -215,18 +236,44 @@ class ResourceIds {
+  /// A double to track the total capacity of the resource, since the whole_ids_ vector
+  /// keeps changing
+  double total_capacity_;
+  /// A double to track any pending decrements in capacity that weren't executed because
```
this is not a double

robertnishihara commented on src/ray/raylet/scheduling_resources.cc:
```diff
@@ -241,9 +307,18 @@ ResourceIds ResourceIds::Acquire(double resource_quantity) {
-  whole_ids_.insert(whole_ids_.end(), whole_ids_to_return.begin(),
-                    whole_ids_to_return.end());
+  int64_t return_resource_count = whole_ids_to_return.size();
+  if (return_resource_count > decrement_backlog_) {
```
We're only clearing the `decrement_backlog_` when we return whole IDs. So if each task only uses fractional resources, then we will never clear the `decrement_backlog_`, right?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13631/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13809/
Test FAILed.

raulchen commented:
@romilbhardwaj @robertnishihara Is this PR ready for merge now? And what's status of the 3rd PR? thanks.

robertnishihara commented:
@raulchen I don't think it is ready. We should merge https://github.com/ray-project/ray/pull/4533 first since there will be conflicts. cc @williamma12

raulchen commented:
@robertnishihara Thanks. In case #4533 still takes long time to merge, I think we can also merge this PR first. Because this PR doesn't introduce new problems.

robertnishihara commented:
#4533 has been merged, so I think this can be rebased and merged.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13973/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13984/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13995/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/13998/
Test FAILed.
